### ISSUE TITLE
Unexpected return type error for overloaded function using enum and if-else clause 

### ISSUE THREAD

**Issue by rpmcginty** [python/mypy#16682]
## **Bug Report**

Using the `overload` decorator and an input enum argument, I am trying to better define the return type for a function. This however is resulting in very unexpected errors related to return types. It seems that mypy does not follow the correct if-else branch based on the enum type argument passed into the function.

### **To Reproduce**

I've included the isolated real-world example as well as a simplified example. Both use an if-else clause and an input argument of type enum to determine the return type of the function. The two examples result in different mypy errors, but I believe they are related to the same issue.

**Real-world example:**
Gist URL: https://gist.github.com/mypy-play/e1541ddeb19e3db03dff7b04860dbc92
Playground URL: https://mypy-play.net/?mypy=1.7.1&python=3.11&gist=e1541ddeb19e3db03dff7b04860dbc92

**Simplified example (different error):**
Gist URL: https://gist.github.com/mypy-play/6deb224c3cb985659dea8c3e9d0b2630
Playground URL: https://mypy-play.net/?mypy=latest&python=3.11&gist=6deb224c3cb985659dea8c3e9d0b2630

```python
class TargetFormat(Enum):
    STR = "str"
    INT = "int"

ItemType = TypeVar("ItemType", str, int)
ItemCollection = TypeVar(
    "ItemCollection",
    Sequence[str],
    Sequence[int],
)


@overload
def convert(
    item_collection: ItemCollection, target_format: Literal[TargetFormat.STR]
) -> List[str]:
    ...


@overload
def convert(
    item_collection: ItemCollection, target_format: Literal[TargetFormat.INT]
) -> List[int]:
    ...


def convert(
    item_collection: ItemCollection, target_format: TargetFormat
) -> List[ItemType]:
    if target_format == TargetFormat.STR:
        return [str(_) for _ in item_collection]
    elif target_format == TargetFormat.INT:
        return [int(_) for _ in item_collection]
    else:
        raise ValueError(f"Invalid collection: {target_format}")

```

### **Expected Behavior**

The following overloads 

### **Actual Behavior**

Mypy fails with a bizarre error. And there are different errors depending on my real-world example and simplified example. But they both fail where the if-else clause using the input enum type argument is evaluated. 

In my real-world example,
```
main.py:158: error: Incompatible return value type (got "list[EnvVarItem]", expected "list[tuple[str, str]]")  [return-value]
main.py:158: error: Incompatible return value type (got "list[EnvVarItem]", expected "list[EnvVarDictItemUpper]")  [return-value]
main.py:158: error: Incompatible return value type (got "list[EnvVarItem]", expected "list[EnvVarDictItemLower]")  [return-value]
Found 3 errors in 1 file (checked 1 source file)
```

In my simplified example, 
```
main.py:51: error: List comprehension has incompatible type List[str]; expected List[int]  [misc]
main.py:53: error: List comprehension has incompatible type List[int]; expected List[str]  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

Interestingly, even if I switch the enums in the if-else clause, the same error is raised by mypy 

### **Your Environment**

- Mypy version used: 1.7.1, 1.5.0, 0.991
- Python version used: 3.9 - 3.12 (see mypy playground)

**Comment by erictraut** [comment]
You're using a TypeVar in an invalid way here. The proper return type for `convert` should be `List[str] | List[int]`. If you fix that problem in the above code sample, it type checks fine.

**Comment by rpmcginty** [comment]
Okay, I see. Using `Union` over `TypeVar` solves the errors in mypy.

there is one more scenario that seems to be problematic. 
If I define a function which subsequently calls the `convert` function, I get an call overload error. 

```python
def some_redirect(
    item_collection: ItemCollection, target_format: TargetFormat
) -> ItemCollection:
    
    # Does something else
    # ...
    return convert(item_collection, target_format)

# main.py:64: error: No overload variant of "convert" matches argument types "Sequence[str] | Sequence[int]", "TargetFormat"  [call-overload]
# main.py:64: note: Possible overload variants:
# main.py:64: note:     def convert(item_collection: Sequence[str] | Sequence[int], target_format: Literal[TargetFormat.STR]) -> list[str]
# main.py:64: note:     def convert(item_collection: Sequence[str] | Sequence[int], target_format: Literal[TargetFormat.INT]) -> list[int]
# Found 1 error in 1 file (checked 1 source file)

```
If I remove the enum variable from the `convert` call, it works. Why is it that the enum variable causes issues for mypy when trying to resolve proper overload method?

**Comment by erictraut** [comment]
> Why is it that the enum variable causes issues for mypy when trying to resolve proper overload method?

There are two issues here.

First, both mypy and pyright do "union expansion" for argument types. That is, if the argument evaluates to a union type, the type checker allows different subtypes of the union to match different overloads. However, neither mypy nor pyright do the equivalent of union expansion for non-union types like an enum. If you want this to type check, you would need to manually expand the union (i.e. replace the type annotation `TargetFormat` with `Literal[TargetFormat.INT, TargetFormat.STR]`).

Second, you're using the `ItemCollection` TypeVar incorrectly. I think you meant for this to be a union, not a TypeVar. Here's a quick tip about type variables: If you are using a TypeVar only once within a function signature, you're probably doing something wrong. A TypeVar should appear at least twice in a function signature.

Here's a fully-corrected code sample that type checks without errors. Note that no TypeVars are involved.

```python
from enum import Enum
from typing import Literal, Sequence, overload

class TargetFormat(Enum):
    STR = "str"
    INT = "int"

ItemCollection = Sequence[str] | Sequence[int]

@overload
def convert(
    item_collection: ItemCollection, target_format: Literal[TargetFormat.STR]
) -> list[str]:
    ...

@overload
def convert(
    item_collection: ItemCollection, target_format: Literal[TargetFormat.INT]
) -> list[int]:
    ...

def convert(
    item_collection: ItemCollection, target_format: TargetFormat
) -> list[str] | list[int]:
    if target_format == TargetFormat.STR:
        return [str(_) for _ in item_collection]
    if target_format == TargetFormat.INT:
        return [int(_) for _ in item_collection]

    raise ValueError(f"Invalid collection: {target_format}")

def some_redirect(
    item_collection: ItemCollection,
    target_format: Literal[TargetFormat.STR, TargetFormat.INT],
) -> ItemCollection:
    return convert(item_collection, target_format)
```

**Comment by rpmcginty** [comment]
> First, both mypy and pyright do "union expansion" for argument types. That is, if the argument evaluates to a union type, the type checker allows different subtypes of the union to match different overloads. However, neither mypy nor pyright do the equivalent of union expansion for non-union types like an enum. If you want this to type check, you would need to manually expand the union (i.e. replace the type annotation TargetFormat with `Literal[TargetFormat.INT, TargetFormat.STR]`).

Ahh, I see now. thanks for the explanation!

> Second, you're using the `ItemCollection` TypeVar incorrectly. I think you meant for this to be a union, not a TypeVar. Here's a quick tip about type variables: If you are using a TypeVar only once within a function signature, you're probably doing something wrong. A TypeVar should appear at least twice in a function signature.

I had actually stopped using `TypeVar` all together when you originally suggested that. sorry for not making that clear.
